File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int goodNodes(TreeNode* root) {
        // 1. premax[&TreeNode] is the max node so far of each node
        // 2. choose if curr > premax[curr]
        // 3. update premax[next]

        unordered_map<TreeNode*, int> premax;
        queue<TreeNode*> q;
        int good = 0;

        if (root) {
            q.push(root);
            premax[root] = root->val;
        }

        while (!q.empty()) {
            TreeNode *curr = q.front();
            q.pop();
            if (premax[curr] <= curr->val) {
                good++;
            }
            if (curr->left) {
                premax[curr->left] = max(premax[curr], curr->val);
                q.push(curr->left);
            }
            if (curr->right) {
                premax[curr->right] = max(premax[curr], curr->val);
                q.push(curr->right);
            }
        }

        return good;
    }
};

File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Codec:
    
    # Encodes a tree to a single string.
    def serialize(self, root: Optional[TreeNode]) -> str:
        # 1,null,3,4,null,null,5,null,null,
        if not root: return "n,"
        return str(root.val) + "," + self.serialize(root.left) + self.serialize(root.right)

    # Decodes your encoded data to tree.
    def deserialize(self, data: str) -> Optional[TreeNode]:
        st, en = 0, 0

        def dfs() -> Optional[TreeNode]:
            nonlocal st, en, data
            while data[en] != ",": en += 1

            s = data[st:en]
            en += 1
            st = en

            if (s == "n"): return
            else: s = int(s)

            curr = TreeNode(s)
            curr.left = dfs()
            curr.right = dfs()
            
            return curr
        
        return dfs()
        

